### PR TITLE
add MSA and GLA operators

### PR DESCRIPTION
**Pull request by czy1101**
# MiniMax M3 Paged Sparse Attention Benchmark Results

## Test Environment

```text
benchmark/minimax_sparse_attention_benchmark.py::test_msa_benchmark

Device: NVIDIA H100 80GB HBM3
Compute Capability: 9.0

Baseline: vLLM enabled
Timing: eager execution with CUDA events
```

------
### Configuration

```text
MiniMax M3 paged sparse attention (prefill)

dtype=bf16
topk=16
init/local=1/2
pages=random
seed=0
warmup=200ms
rep=300ms

Provider order: alternates by shape
```

### Benchmark Results

| Shape `[B,S,KVH,H]` | flag_attn (ms) | vLLM (ms) | vLLM / ours | Score (ms) | TopK (ms) | Attn (ms) |
| ------------------- | ------------- | --------- | ----------- | ---------- | --------- | --------- |
| 1×8192×16×96        | 12.8530       | 16.1406   | **1.26×**   | 0.3710     | 0.1238    | 12.3712   |
| 2×16384×8×96        | 31.5028       | 37.9965   | **1.21×**   | 1.3258     | 0.6276    | 29.5778   |
| 1×32768×16×96       | 68.8183       | 79.2246   | **1.15×**   | 5.1454     | 3.1332    | 60.8095   |
| 2×8192×8×96         | 13.5406       | 17.1950   | **1.27×**   | 0.3718     | 0.1236    | 13.0513   |
| 4×4096×16×384       | 28.3438       | 41.3379   | **1.46×**   | 0.3720     | 0.1761    | 27.7693   |
| 4×4096×16×256       | 21.5160       | 28.9681   | **1.35×**   | 0.3711     | 0.1762    | 20.9829   |

------
### Benchmark Results

| Shape `[B,S,KVH,H]` | flag_attn(ms) | vLLM (ms) | vLLM / ours | Score (ms) | TopK (ms) | Attn (ms) |
| ------------------- | ------------- | --------- | ----------- | ---------- | --------- | --------- |
| 1×8192×16×96        | 17.3029       | 26.2415   | **1.52×**   | 0.2023     | 0.1236    | 16.9745   |
| 2×16384×8×96        | 42.7100       | 71.0890   | **1.66×**   | 0.7141     | 0.6323    | 41.3461   |
| 1×32768×16×96       | 79.8781       | 117.9459  | **1.48×**   | 2.7282     | 3.1327    | 74.1841   |
| 2×8192×8×96         | 19.8213       | 33.2899   | **1.68×**   | 0.2025     | 0.1236    | 19.4617   |
| 4×4096×16×384       | 49.4060       | 80.0902   | **1.62×**   | 0.1941     | 0.1761    | 48.9741   |
| 4×4096×16×256       | 34.3683       | 57.3066   | **1.67×**   | 0.1941     | 0.1761    | 33.9236   |

------

### Benchmark Results

| Shape `[B,S,KVH,H]` | flag_attn(ms) | vLLM (ms) | vLLM / ours | IdxDec (ms) | AttnDec (ms) |
| ------------------- | ------------- | --------- | ----------- | ----------- | ------------ |
| 1×4096×16×96        | 0.2225        | 0.2102    | **0.94×**   | 0.1043      | 0.0428       |
| 1×16384×16×96       | 0.2104        | 0.2090    | **0.99×**   | 0.1048      | 0.0420       |
| 1×65536×16×96       | 0.2115        | 0.2091    | **0.99×**   | 0.1066      | 0.0422       |
| 4×4096×8×96         | 0.2133        | 0.2117    | **0.99×**   | 0.1027      | 0.0430       |
| 4×16384×8×96        | 0.2128        | 0.2111    | **0.99×**   | 0.1044      | 0.0427       |
| 16×4096×8×96        | 0.2166        | 0.2178    | **1.01×**   | 0.1029      | 0.0644       |
| 32×2048×4×48        | 0.0870        | 0.2204    | **2.53×**   | 0.0055      | 0.0641       |
| 64×1024×4×48        | 0.0878        | 0.2177    | **2.48×**   | 0.0059      | 0.0660       |

------

### Benchmark Results

| Shape `[B,S,KVH,H]` | flag_attn(ms) | vLLM (ms) | vLLM / ours | IdxDec (ms) | AttnDec (ms) |
| ------------------- | ------------- | --------- | ----------- | ----------- | ------------ |
| 1×4096×16×96        | 0.2164        | 0.2124    | **0.98×**   | 0.1025      | 0.0443       |
| 1×16384×16×96       | 0.2099        | 0.2100    | **1.00×**   | 0.1033      | 0.0433       |
| 1×65536×16×96       | 0.2113        | 0.2105    | **1.00×**   | 0.1032      | 0.0437       |
| 4×4096×8×96         | 0.2215        | 0.2148    | **0.97×**   | 0.1053      | 0.0454       |
| 4×16384×8×96        | 0.2175        | 0.2186    | **1.01×**   | 0.1056      | 0.0455       |
| 16×4096×8×96        | 0.2221        | 0.2268    | **1.02×**   | 0.1032      | 0.0595       |
| 32×2048×4×48        | 0.0900        | 0.2268    | **2.52×**   | 0.0056      | 0.0597       |
| 64×1024×4×48        | 0.0942        | 0.2238    | **2.38×**   | 0.0056      | 0.0617       |

------

## Test Status

```text
PASSED
```

Note: When topk=32, the decode stage can achieve better performance.

```python
pytest -vv -s tests/flag_attn/test_minimax_sparse_attention.py 

pytest -vv -s benchmark/minimax_sparse_attention_benchmark.py 

```

### CODE REVIEW

File: benchmark/chunk_gla_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
+#!/usr/bin/env python3
```
Comment (huangyiqun) [inline benchmark/chunk_gla_benchmark.py]: Plz remove this line.

File: benchmark/chunk_gla_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Forward and backward benchmark for chunked gated linear attention."""
+
+from __future__ import annotations
```
Comment (huangyiqun) [inline benchmark/chunk_gla_benchmark.py]: Will this affect the performance test results?

File: benchmark/chunk_gla_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    from fla.ops.gla import chunk_gla as _fla_chunk_gla
+
+    _HAS_FLA_CHUNK = True
+except Exception:
+    _HAS_FLA_CHUNK = False
+
+
+def _fla_chunk_wrapper(q, k, v, g, **kwargs):
+    if not _HAS_FLA_CHUNK:
+        raise RuntimeError("fla chunk_gla is unavailable")
```
Comment (huangyiqun) [inline benchmark/chunk_gla_benchmark.py]: Don't raise a RuntimeError here. If fla is unavailable, just skip certain tests and print the relevant information.

File: benchmark/chunk_gla_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+        print(
+            f"{B:>3} {T:>6} {H:>4} {D:>4} {dtype_str:>8} "
+            f"{ms_flag_attn:>14.3f}"
+        )
+
+
+# ---------------------------
+# unified benchmark (fwd then fwd+bwd)
+# ---------------------------
+def run_benchmark(warmup: int = DEFAULT_WARMUP, rep: int = DEFAULT_REP) -> None:
```
Comment (huangyiqun) [inline benchmark/chunk_gla_benchmark.py]: Use pytest for the accuracy and performance tests of the newly submitted operator.

File: benchmark/minimax_sparse_attention_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,807 @@
+#!/usr/bin/env python3
```
Comment (huangyiqun) [inline benchmark/minimax_sparse_attention_benchmark.py]: Plz remove this line.

File: benchmark/minimax_sparse_attention_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+
+    for mode_index, decode in enumerate(modes):
+        args.decode = decode
+        if mode_index:
+            print()
+        for dtype_name in dtypes:
+            _run_dtype(args, dtype_name)
+
+
+def test_msa_benchmark(request) -> None:
```
Comment (huangyiqun) [inline benchmark/minimax_sparse_attention_benchmark.py]: Just like the comments in `chunk_gla_benchmark.py`, please use pytest for the test functions.

File: src/flag_attn/gated_linear_attention/compat.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+"""Compatibility helpers for running the GLA kernels without FlagGems."""
+
+from collections.abc import Sequence
+from typing import Any
+
+import triton
+
+
+def libtuner(
```
Comment (huangyiqun) [inline src/flag_attn/gated_linear_attention/compat.py:23]: What is the difference between this and Triton's autotune?

File: src/flag_attn/minimax_sparse_attention/index_topk.py [hunk 8]
```diff
@@ -0,0 +1,1617 @@
...
+    else:
+        _topk_index_kernel_fallback[grid_topk](
+            *kernel_args,
+            MASK_INIT=False,
+            MASK_LOCAL=False,
+        )
+    return topk_idx
+
+
+@torch.no_grad()
```
Comment (huangyiqun) [inline src/flag_attn/minimax_sparse_attention/index_topk.py:1342]: Is this necessary?

File: src/flag_attn/gated_linear_attention/index.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+def prepare_chunk_indices(
+    cu_seqlens: torch.LongTensor,
+    chunk_size: int,
+    cu_seqlens_cpu: torch.LongTensor | None = None,
+) -> torch.LongTensor:
+    # The FlagGems GLA wrapper forwards this optional host mirror for API
+    # compatibility; index construction only needs the device tensor.
+    del cu_seqlens_cpu
+    chunk_counts = triton.cdiv(prepare_lens(cu_seqlens), chunk_size)
+    chunk_offsets = torch.cat([cu_seqlens.new_tensor([0]), chunk_counts]).cumsum(-1)
```
Comment (huangyiqun) [inline src/flag_attn/gated_linear_attention/index.py:27]: This doesn't seem to use a Triton kernel. Can it be fused into another Triton kernel?

File: src/flag_attn/gated_linear_attention/index.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    chunk_arange = torch.arange(chunk_offsets[-1], device=cu_seqlens.device)
```
Comment (huangyiqun) [inline src/flag_attn/gated_linear_attention/index.py:28]: and this

File: src/flag_attn/gated_linear_attention/index.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    seq_ids = torch.repeat_interleave(
+        torch.arange(chunk_counts.numel(), device=cu_seqlens.device),
```
Comment (huangyiqun) [inline src/flag_attn/gated_linear_attention/index.py:30]: and this

File: src/flag_attn/gated_linear_attention/index.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        chunk_counts,
+    )
+    chunk_ids = chunk_arange - torch.repeat_interleave(chunk_offsets[:-1], chunk_counts)
+    return torch.stack([seq_ids, chunk_ids], 1).to(cu_seqlens)
```
Comment (huangyiqun) [inline src/flag_attn/gated_linear_attention/index.py:34]: this

File: src/flag_attn/minimax_sparse_attention/utils.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+current_platform = _CurrentPlatform()
+
+
+def round_up(n: int, d: int) -> int:
+    """Round n up to the nearest multiple of d."""
+    return (n + d - 1) // d * d
+
+
+def has_triton_tle(major: int = 0, minor: int = 0, patch: int = 0) -> bool:
```
Comment (huangyiqun) [inline src/flag_attn/minimax_sparse_attention/utils.py:48]: Can this function be extracted as a common implementation for future use?

File: tests/flag_attn/test_chunk_gla.py
Comment (huangyiqun) [inline tests/flag_attn/test_chunk_gla.py:1]: Plz add `@pytest.mark.{op_name}` before test functions.

File: tests/flag_attn/test_minimax_sparse_attention.py [hunk 14]
```diff
@@ -0,0 +1,826 @@
+#!/usr/bin/env python3
```
Comment (huangyiqun) [inline tests/flag_attn/test_minimax_sparse_attention.py]: Plz remove this line.

File: tests/flag_attn/test_minimax_sparse_attention.py
Comment (huangyiqun) [inline tests/flag_attn/test_minimax_sparse_attention.py:1]: Plz add @pytest.mark.{op_name} before test functions.

File: benchmark/chunk_gla_benchmark.py (hunk 2 above)
Comment (czy1101) [reply benchmark/chunk_gla_benchmark.py]: Removed

File: benchmark/chunk_gla_benchmark.py (hunk 3 above)
Comment (czy1101) [reply benchmark/chunk_gla_benchmark.py]: Fixed

File: src/flag_attn/gated_linear_attention/compat.py (hunk 7 above)
Comment (czy1101) [reply src/flag_attn/gated_linear_attention/compat.py:23]: It is essentially autotuning across different input shapes and providers.

File: src/flag_attn/minimax_sparse_attention/index_topk.py (hunk 8 above)
Comment (czy1101) [reply src/flag_attn/minimax_sparse_attention/index_topk.py:1342]: It follows the same usage as vLLM, and backward is not considered here.

File: src/flag_attn/gated_linear_attention/index.py (hunk 9 above)
Comment (czy1101) [reply src/flag_attn/gated_linear_attention/index.py:27]: This is a reusable chunking function used inside kernels by different operators.

**Review by huangyiqun — APPROVED** [review]
